### ISSUE TITLE
Oidc-Client-Filter: Retry with refreshed token on authorization error

### ISSUE THREAD

**Issue by fwippe** [quarkusio/quarkus#16557]
## Description

The OIDC Client Filter (`OidcClientRequestFilter`) should retry failed requests with a refreshed access token. Retries must only be done on authorization errors, maybe really only on return status `401`, and only once.

## Implementation ideas

We could introduce a new `RequestScoped` bean `RetryIndicator` with a property `retrying`. A new `@Interceptor` could wrap the REST client calls and catch arising `WebApplicationException`s with status `401`. The interceptor could then set the `RetryIndicator.retrying` property and retry the invocation. The `OidcClientRequestFilter` could inject the `RetryIndicator` and ask for the `retrying` state. When `true` it would use a refreshed access token.

/cc @sberyozkin What do you think?

**Comment by sberyozkin** [comment]
@fwippe thanks for the idea;
I was thinking about it but I really did not want it going beyond its role of acquiring the tokens - dealing with the failures related to the actual use of tokens is out of scope for what it has been created for which is to facilitate the token acquisition.
The only safe retry is when the token has expired - the filter has no idea if the token's signature has been broken in transit /etc, So preventing 401 related to the expiry is the only thing it can do - and we already have a solution for it - the time skew for refreshing the tokens - setting it say 10 sec will ensure 401 related to the expiry will never happen
As such, IMHO, this issue should not be implemented

**Comment by fwippe** [comment]
@sberyozkin 

We've been using the request filter for a few weeks now (using a time skew of 5 minutes). Sometimes the filter seems stuck in a state where it would simply reuse the token every time, never refreshing the token. I was not yet able to reproduce this state reliably, but I suppose there might be a race condition in the CAS Code of the `TokensHelper`. I was hoping this MR might offer a last resort for such stuck states. Well, I suppose such errors should be fixed at the source. Feel free to close the issue.

**Comment by sberyozkin** [comment]
@fwippe Hi, OK, I see.

I agree with

>  I suppose such errors should be fixed at the source (even though concurrency issues are hard to find and even harder to fix)

big +1 :-)

If we get 401 - it could be N reasons - so even if there were no concerns about the race condition then attempting to refresh in all of those N-1 (which is exp issue) would only stress the system unnecessarily.

Please open an issue to investigate the CAS code.
